core/src/main/java/org/apache/carbondata/core/datamap/DataMapUtil.java

```diff
@@ -145,18 +147,23 @@ public static DataMapJob getEmbeddedJob() {
    * Prune the segments from the already pruned blocklets.
    */
   public static void pruneSegments(List<Segment> segments, List<ExtendedBlocklet> prunedBlocklets) {
-    Set<Segment> validSegments = new HashSet<>();
+    Map<Segment, Set<String>> validSegments = new HashMap<>();
     for (ExtendedBlocklet blocklet : prunedBlocklets) {
-      // Clear the old pruned index files if any present
-      blocklet.getSegment().getFilteredIndexShardNames().clear();
       // Set the pruned index file to the segment
       // for further pruning.
       String shardName = CarbonTablePath.getShardName(blocklet.getFilePath());
-      blocklet.getSegment().setFilteredIndexShardName(shardName);
-      validSegments.add(blocklet.getSegment());
+      // Add the existing shards to corresponding segments
+      Set<String> existingShards = new HashSet<>();
+      existingShards = validSegments.putIfAbsent(blocklet.getSegment(), existingShards);
```

Comment:
Change this to getOrDefault so it will avoid creating empty set every time

Reply:
done